### ISSUE TITLE
High Resolution Layer - Grassland

### ISSUE THREAD

**Issue by Mohinem** [FAIRiCUBE/data-requests#26]

**Comment by KathiSchleidt** [comment]
Same as #10, setting to provisioning here

**Comment by robknapen** [comment]
I have similar remarks for this dataset as for the Corine dataset: Choose a better label for the temporal dimension than 'ansi', and the Envelope coordinates need to be corrected.

Also, Copernicus lists the temporal extent as (the year) 2018. I don't think it is correct to reduce this to a temporal extent of only the first day of 2018, it should cover the full year.

**Comment by Mohinem** [comment]
@robknapen I didn't understand your concern about the Envelope coordinates (infact I don't know what it means). I would like you to explain.

**Comment by robknapen** [comment]
Hi @Mohinem , the Envelope, in the OGC definition, is the minimum bounding box of the coverage., and part of the metadata of the dataset. Such envelopes and bounding boxes are often used in (spatial) searching.

I expect it is also used to indicate the footprints of the geo-referenced dataset in the rasdaman ows (see screenshot). But that does not match the data then, since the grassland dataset only covers Europe (and not the ocean etc.). So some correction is needed I think.

![Screenshot 2023-06-23 at 11 48 02](https://github.com/FAIRiCUBE/data-requests/assets/1943289/2dd90502-df9a-43fb-9894-969cb24c4f2e)

**Comment by Mohinem** [comment]
Ok, so it's bad because it doesn't cover southern Europe ? Like Italy, Spain, Portugal ?

**Comment by robknapen** [comment]
Indeed. So the WGS84 bounding box coordinates for this coverage are not set correctly.

![Screenshot 2023-06-23 at 12 17 09](https://github.com/FAIRiCUBE/data-requests/assets/1943289/963b2fb6-83b4-4296-9cd1-8cf43299316f)

If you look at the CRS for the dataset (on https://epsg.io/3035), it basically also gives you (below the small map on the right) the **maximum** bounding box (with the WGS84 bounds as well). So the **minimum** bounding box of a dataset using this projection can be expected to be within this maximum bounding box.

**Comment by KathiSchleidt** [comment]
Hi Both, 
this is an issue with how the rasdaman GUI portrays the 4326 bbox, drawing the lower line of the bbox straight, not following the meridians correctly

The other issue is that the LAEA grid axes are not aligned with those from 4326, leading to very different bboxes. Here a view with the LAEA grid extent in blue, the surrounding 4326 bbox in red:

![CopernicusBBwWGS84BBox](https://github.com/FAIRiCUBE/data-requests/assets/11915304/6da4bc54-6ed5-474e-815b-5b6c44f1da1a)

When I change the map projection, I end up with the same error as shows up in the rasdaman GUI:

![CopernicusBBwWGS84BBox_LAEA](https://github.com/FAIRiCUBE/data-requests/assets/11915304/12113f15-f19a-454f-915e-b371cd01fd3a)

TL;DR; the WGS84 bbox in rasdaman is correct from the coordinates, just displayed wrong at the lower edge. The surface area of the WGS84 bbox is quite a bit bigger than the surface covered by the LAEA Grid.

**Comment by Mohinem** [comment]
I just received the unfortunate news from my team that changing the axis of an already imported coverage is impossible, so I might have to reimport again 😞 .... This time I want to keep it and reimport a small dataset with another name, ask the UC to verify, and then go ahead. What are your thoughts about it ?

**Comment by robknapen** [comment]
For large datasets it sounds like a good approach to start with a sample and have that verified first, before ingesting the full dataset.

**Comment by Mohinem** [comment]
The smaller dataset is called grassland_status_demo, please check if it satisfies the requirements.

**Comment by KathiSchleidt** [comment]
@Mohinem first thing I notice is that in contrast to the other coverages (e.g. corine_land_cover), you've now started using gmlrgrid:ReferenceableGridByVectors in place of gml:RectifiedGrid for the DomainSet. What's the reason for this? Axes are the same.

This also seems to be impacting on the GUI, whereas for corine_land_cover I see a resolution of 1 on the time axis, under grassland_status_demo we just get N/A

**Comment by robknapen** [comment]
I don't have the Copernicus grassland 2018 dataset so cannot compare on content at the moment. But I am still wondering if we indeed want the time axis of this dataset to be in days with Jan 1, 2018 the only value? The documentation of the dataset mentions it as reference year 2018 (March-October).

Also, the dataset seems to have 240 as null value. I don't see that in the Copernicus documentation (they only specify possible values 0, 1, 254 and 255)? And another weird issue is that when I extract a small slice in CSV format it contains those values 240, but in the same slice extracted in JSON format the values are nulls. Why are these different?

**Comment by Mohinem** [comment]
@robknapen Alright, so I used ansidate CRS, which requires month and day. I didn't specify any month and day myself, so it looks like it's a default value. If it's very important to have only the year, I will look at some other CRS for time.

As for 240 as null value, I didn't specify it, but it looks like it's present in the TIFF files of the dataset as a nodata value (can be seen with gdalinfo command).
`mbasak@fairicube:~/grassland/status/GRA_2018_010m_al_03035_v010/DATA$ gdalinfo GRA_2018_010m_E50N21_03035_v010.tif
Driver: GTiff/GeoTIFF
Files: GRA_2018_010m_E50N21_03035_v010.tif
Size is 10000, 10000
Coordinate System is:
PROJCRS["ETRS_1989_LAEA",
    BASEGEOGCRS["ETRS89",
        DATUM["European Terrestrial Reference System 1989",
            ELLIPSOID["GRS 1980",6378137,298.257222101004,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4258]],
    CONVERSION["Lambert Azimuthal Equal Area",
        METHOD["Lambert Azimuthal Equal Area",
            ID["EPSG",9820]],
        PARAMETER["Latitude of natural origin",52,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",10,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["False easting",4321000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",3210000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["(N)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    ID["EPSG",3035]]
Data axis to CRS axis mapping: 1,2
Origin = (5000000.000000000000000,2200000.000000000000000)
Pixel Size = (10.000000000000000,-10.000000000000000)
Metadata:
  AREA_OR_POINT=Area
  DataType=Thematic
Image Structure Metadata:
  COMPRESSION=LZW
  INTERLEAVE=BAND
Corner Coordinates:
Upper Left  ( 5000000.000, 2200000.000) ( 18d15'43.07"E, 42d34'12.20"N)
Lower Left  ( 5000000.000, 2100000.000) ( 18d 8'24.49"E, 41d40'21.46"N)
Upper Right ( 5100000.000, 2200000.000) ( 19d28' 7.52"E, 42d27'50.13"N)
Lower Right ( 5100000.000, 2100000.000) ( 19d19'46.50"E, 41d34' 5.89"N)
Center      ( 5050000.000, 2150000.000) ( 18d47'59.71"E, 42d 4'14.29"N)
Band 1 Block=256x256 Type=Byte, ColorInterp=Palette
  Min=0.000 Max=255.000 
  Minimum=0.000, Maximum=255.000, Mean=172.155, StdDev=119.375
  NoData Value=240
  Metadata:
    BandName=Band_1
    RepresentationType=THEMATIC
    STATISTICS_MAXIMUM=255
    STATISTICS_MEAN=172.15516149
    STATISTICS_MINIMUM=0
    STATISTICS_SKIPFACTORX=1
    STATISTICS_SKIPFACTORY=1
    STATISTICS_STDDEV=119.37507142803
  Color Table (RGB with 256 entries)
    0: 240,240,240,255
    1: 163,199,56,255
    2: 0,0,0,255
    3: 0,0,0,255
    4: 0,0,0,255
    5: 0,0,0,255
    6: 0,0,0,255
    7: 0,0,0,255
    8: 0,0,0,255
    9: 0,0,0,255
   10: 0,0,0,255
   11: 0,0,0,255
   12: 0,0,0,255
   13: 0,0,0,255
   14: 0,0,0,255
   15: 0,0,0,255
   16: 0,0,0,255
   17: 0,0,0,255
   18: 0,0,0,255
   19: 0,0,0,255
   20: 0,0,0,255
   21: 0,0,0,255
   22: 0,0,0,255
   23: 0,0,0,255
   24: 0,0,0,255
   25: 0,0,0,255
   26: 0,0,0,255
   27: 0,0,0,255
   28: 0,0,0,255
   29: 0,0,0,255
   30: 0,0,0,255
   31: 0,0,0,255
   32: 0,0,0,255
   33: 0,0,0,255
   34: 0,0,0,255
   35: 0,0,0,255
   36: 0,0,0,255
   37: 0,0,0,255
   38: 0,0,0,255
   39: 0,0,0,255
   40: 0,0,0,255
   41: 0,0,0,255
   42: 0,0,0,255
   43: 0,0,0,255
   44: 0,0,0,255
   45: 0,0,0,255
   46: 0,0,0,255
   47: 0,0,0,255
   48: 0,0,0,255
   49: 0,0,0,255
   50: 0,0,0,255
   51: 0,0,0,255
   52: 0,0,0,255
   53: 0,0,0,255
   54: 0,0,0,255
   55: 0,0,0,255
   56: 0,0,0,255
   57: 0,0,0,255
   58: 0,0,0,255
   59: 0,0,0,255
   60: 0,0,0,255
   61: 0,0,0,255
   62: 0,0,0,255
   63: 0,0,0,255
   64: 0,0,0,255
   65: 0,0,0,255
   66: 0,0,0,255
   67: 0,0,0,255
   68: 0,0,0,255
   69: 0,0,0,255
   70: 0,0,0,255
   71: 0,0,0,255
   72: 0,0,0,255
   73: 0,0,0,255
   74: 0,0,0,255
   75: 0,0,0,255
   76: 0,0,0,255
   77: 0,0,0,255
   78: 0,0,0,255
   79: 0,0,0,255
   80: 0,0,0,255
   81: 0,0,0,255
   82: 0,0,0,255
   83: 0,0,0,255
   84: 0,0,0,255
   85: 0,0,0,255
   86: 0,0,0,255
   87: 0,0,0,255
   88: 0,0,0,255
   89: 0,0,0,255
   90: 0,0,0,255
   91: 0,0,0,255
   92: 0,0,0,255
   93: 0,0,0,255
   94: 0,0,0,255
   95: 0,0,0,255
   96: 0,0,0,255
   97: 0,0,0,255
   98: 0,0,0,255
   99: 0,0,0,255
  100: 0,0,0,255
  101: 0,0,0,255
  102: 0,0,0,255
  103: 0,0,0,255
  104: 0,0,0,255
  105: 0,0,0,255
  106: 0,0,0,255
  107: 0,0,0,255
  108: 0,0,0,255
  109: 0,0,0,255
  110: 0,0,0,255
  111: 0,0,0,255
  112: 0,0,0,255
  113: 0,0,0,255
  114: 0,0,0,255
  115: 0,0,0,255
  116: 0,0,0,255
  117: 0,0,0,255
  118: 0,0,0,255
  119: 0,0,0,255
  120: 0,0,0,255
  121: 0,0,0,255
  122: 0,0,0,255
  123: 0,0,0,255
  124: 0,0,0,255
  125: 0,0,0,255
  126: 0,0,0,255
  127: 0,0,0,255
  128: 0,0,0,255
  129: 0,0,0,255
  130: 0,0,0,255
  131: 0,0,0,255
  132: 0,0,0,255
  133: 0,0,0,255
  134: 0,0,0,255
  135: 0,0,0,255
  136: 0,0,0,255
  137: 0,0,0,255
  138: 0,0,0,255
  139: 0,0,0,255
  140: 0,0,0,255
  141: 0,0,0,255
  142: 0,0,0,255
  143: 0,0,0,255
  144: 0,0,0,255
  145: 0,0,0,255
  146: 0,0,0,255
  147: 0,0,0,255
  148: 0,0,0,255
  149: 0,0,0,255
  150: 0,0,0,255
  151: 0,0,0,255
  152: 0,0,0,255
  153: 0,0,0,255
  154: 0,0,0,255
  155: 0,0,0,255
  156: 0,0,0,255
  157: 0,0,0,255
  158: 0,0,0,255
  159: 0,0,0,255
  160: 0,0,0,255
  161: 0,0,0,255
  162: 0,0,0,255
  163: 0,0,0,255
  164: 0,0,0,255
  165: 0,0,0,255
  166: 0,0,0,255
  167: 0,0,0,255
  168: 0,0,0,255
  169: 0,0,0,255
  170: 0,0,0,255
  171: 0,0,0,255
  172: 0,0,0,255
  173: 0,0,0,255
  174: 0,0,0,255
  175: 0,0,0,255
  176: 0,0,0,255
  177: 0,0,0,255
  178: 0,0,0,255
  179: 0,0,0,255
  180: 0,0,0,255
  181: 0,0,0,255
  182: 0,0,0,255
  183: 0,0,0,255
  184: 0,0,0,255
  185: 0,0,0,255
  186: 0,0,0,255
  187: 0,0,0,255
  188: 0,0,0,255
  189: 0,0,0,255
  190: 0,0,0,255
  191: 0,0,0,255
  192: 0,0,0,255
  193: 0,0,0,255
  194: 0,0,0,255
  195: 0,0,0,255
  196: 0,0,0,255
  197: 0,0,0,255
  198: 0,0,0,255
  199: 0,0,0,255
  200: 0,0,0,255
  201: 0,0,0,255
  202: 0,0,0,255
  203: 0,0,0,255
  204: 0,0,0,255
  205: 0,0,0,255
  206: 0,0,0,255
  207: 0,0,0,255
  208: 0,0,0,255
  209: 0,0,0,255
  210: 0,0,0,255
  211: 0,0,0,255
  212: 0,0,0,255
  213: 0,0,0,255
  214: 0,0,0,255
  215: 0,0,0,255
  216: 0,0,0,255
  217: 0,0,0,255
  218: 0,0,0,255
  219: 0,0,0,255
  220: 0,0,0,255
  221: 0,0,0,255
  222: 0,0,0,255
  223: 0,0,0,255
  224: 0,0,0,255
  225: 0,0,0,255
  226: 0,0,0,255
  227: 0,0,0,255
  228: 0,0,0,255
  229: 0,0,0,255
  230: 0,0,0,255
  231: 0,0,0,255
  232: 0,0,0,255
  233: 0,0,0,255
  234: 0,0,0,255
  235: 0,0,0,255
  236: 0,0,0,255
  237: 0,0,0,255
  238: 0,0,0,255
  239: 0,0,0,255
  240: 0,0,0,0
  241: 0,0,0,255
  242: 0,0,0,255
  243: 0,0,0,255
  244: 0,0,0,255
  245: 0,0,0,255
  246: 0,0,0,255
  247: 0,0,0,255
  248: 0,0,0,255
  249: 0,0,0,255
  250: 0,0,0,255
  251: 0,0,0,255
  252: 0,0,0,255
  253: 0,0,0,255
  254: 153,153,153,255
  255: 0,0,0,255
`
As for why slices have different outputs in CSV or JSON, I don't know, will have to find it out.

@KathiSchleidt to answer your both questions-
1. I don't define the DomainSet. It has happened automatically, and I will try finding the reason.
2. I am sure that it's because of using irregular time axis, as I had to specify the resolution for regular_time_axis, but I can't decide if the resolution for 1 year is 365 or 366 days (for leap year), so I thought irregular time axis is a better choice for time axis.

**Comment by Mohinem** [comment]
I realised that it's hard to view the definition of nodata value above, so here it is-
`  NoData Value=240 `

**Comment by KathiSchleidt** [comment]
I've now done a comparison with the original grassland_status dataset, get totally different values for the same subset:

grassland_status request: 
`https://fairicube.rasdaman.com/rasdaman/ows?&SERVICE=WCS&VERSION=2.0.1&REQUEST=GetCoverage&COVERAGEID=grassland_status&SUBSET=ansi("2018-01-01T08:12:02.000Z")&SUBSET=Y(2200000,2200100)&SUBSET=X(5000000,5000100)&FORMAT=application/gml+xml`

Response:
```
<gml:rangeSet>
<gml:DataBlock>
<gml:rangeParameters/>
<gml:tupleList cs=" " ts=",">
0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
</gml:tupleList>
</gml:DataBlock>
</gml:rangeSet>
```

grassland_status_demo request:

`https://fairicube.rasdaman.com/rasdaman/ows?&SERVICE=WCS&VERSION=2.0.1&REQUEST=GetCoverage&COVERAGEID=grassland_status_demo&SUBSET=Time("2018-01-01T00:00:00.000Z")&SUBSET=Y(2200000,2200100)&SUBSET=X(5000000,5000100)&FORMAT=application/gml+xml`

Response: 
```
<gml:rangeSet>
<gml:DataBlock>
<gml:rangeParameters/>
<gml:tupleList cs=" " ts=",">
240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240,240
</gml:tupleList>
</gml:DataBlock>
</gml:rangeSet>
```

In addition, I'm having issues with the requests on both of these coverages. When I curl the request URL from the GUI, it returns an error (there's only one date value in the subset):  <ows:ExceptionText>Trimming subset in WCS must follow this syntax 'lowerBound,upperBound', given '2018-01-01T08:12:02.000Z'.</ows:ExceptionText>

When I add a 2nd date, I get a curl error: Bad Request</h1> <p>Your browser sent a request that this server could not understand. 

When I uuencode the dates as follows, I also get an error:

curl 'https://fairicube.rasdaman.com/rasdaman/ows?&SERVICE=WCS&VERSION=2.0.1&REQUEST=GetCoverage&COVERAGEID=grassland_status&SUBSET=ansi(%222018-01-01T08%3A12%3A02.000Z%22%2C%20%222018-01-01T08%3A12%3A02.000Z%22)&SUBSET=Y(2200000,2200100)&SUBSET=X(5000000,5000100)&FORMAT=application/gml+xml' -u "fairicube_dc:p.aZDghQg0"

<ows:ExceptionText><![CDATA[A parsing error occurred at line '1', column '77'. Offending token is '"2018-01-01T08:12:02.000Z"'. Reason: no viable alternative at input '+"2018-01-01T08:12:02.000Z"'.]]></ows:ExceptionText>

Any insights why this is not working?

**Comment by robknapen** [comment]
I think somehow in the TIFF files the no-data value is used to refer to index 240 in the colour table. So the content and visualisation are perhaps getting mixed up?

**Comment by KathiSchleidt** [comment]
on making this a time series, how do you propose resolving the 10&100m resolution of the 2018 dataset with the 20&100m resolution of the 2015 dataset?

this increasing resolution will be a general issue with the datasets, I see 2 options:
- creating a separate coverage for the 10&20m years
- merging the 10m cells to 20m

**Comment by robknapen** [comment]
It seems to me that the grassland 2015 and grassland 2018 are separate data products, and a change dataset is also being provided. I would keep it that way for now.

**Comment by Mohinem** [comment]
Any modifications required on my part for the demo dataset ?

**Comment by robknapen** [comment]
Do we have solutions for the things that are mentioned before?

**Comment by KathiSchleidt** [comment]
Hi @Mohinem 
I've now created dedicated issues for the 2 I noticed here pertaining to the Grassland dataset:
- [Dimensional structure for datasets](https://github.com/FAIRiCUBE/data-requests/issues/43)
- [Handling irregular series](https://github.com/FAIRiCUBE/data-requests/issues/44)

On the Grassland dataset, what are your proposals?

**Comment by Mohinem** [comment]
For the grassland dataset,
1. Dimension remains same - lat/long/time. In case we want to just keep the year in the 'time', we require a different crs for resolving time compared to ansidate.
2. Two different datasets seem fine (unless there are specific requirements).

**Comment by KathiSchleidt** [comment]
What's the issue with ansidate? Can only do daily (or multiday) series, not yearly?

**Comment by Mohinem** [comment]
Yes. Ansidate requires month and day information, so the month and day was assigned even if I only configured the year.

**Comment by robknapen** [comment]
Perhaps it is a matter of FAIRiCUBE defining how we interpret the temporal axis for data sets such as this. I.e. Copernicus says that this is a dataset from 2018 (collected between March - October). So when we use a time dimension and it must use day resolution, which date do we choose (Jan 1, March 1, Oct 31, Dec 31)?

**Comment by KathiSchleidt** [comment]
Is the issue that ansidate includes day and month information, or that its unclear how to define yearly (or in this case multiyear) intervals?

**Comment by Mohinem** [comment]
Yeah, ansidate format itself requires month and day information, and if I don't provide it assigns a month and day. If there's some time crs for just year and interval, I can use that.

**Comment by pebau** [comment]
Dear all, I find a long, winding discussion here. Please simply provide all information (either directly or via link) on nulls, CRS, extent, envelope, etc. documented in the data request. To void further such circular activities I have advised Mohit to start only once all information is available unambiguously in the data request.

**Comment by Mohinem** [comment]
@robknapen  two questions for you.

1. How do you like the date to be represented ?
 a. Geo-extent ["2018-01-01T00:00:00.000Z", "2018-12-31T23:59:59.999Z"]
 b. Just a year ("2018") ?
Or are both representations fine ?
 
2. is 240 as nodata value fine ? Or should that be 0 ?

**Comment by robknapen** [comment]
Hi @Mohinem,

I would give it a time axis that has year as units, and 2018 as only value. My interpretation would be that the data is 'valid' for that time (e.g. that whole year, no more precise details are known).

For the nodata value I think it is best to keep the 240. Somebody selected that on purpose, probably because the values 0 and 255 were already in use and their definition doesn't match the concept of having no data for a grid cell.

**Comment by pebau** [comment]
> I would give it a time axis that has year as units, and 2018 as only value. My interpretation would be that the data is 'valid' for that time (e.g. that whole year, no more precise details are known).

hm, but you are aware that you create a 3D coverage this way which cannot be used straightfoward as 2D any more? Eg, in WMS always provide the TIME parameter, in WCS need time slicing to encode in 2D TIFF, etc.

**Comment by robknapen** [comment]
Indeed, but what are the alternatives? Include the data validity timespan in data cube metadata somehow in a consistent way? Or would it end-up in the name/ID of the data cube? Is there a best practice in the data cube world?

**Comment by pebau** [comment]
> Indeed, but what are the alternatives? Include the data validity timespan in data cube metadata somehow in a consistent way? Or would it end-up in the name/ID of the data cube? Is there a best practice in the data cube world?

as you already sense, Rob: "propagating up" the time information into the general (catalog) metadata would be a way to go. 

PS: Never ever put semantics into the name!

**Comment by KathiSchleidt** [comment]
A time dimension in years would be valuable across the Copernicus Products. I'd understood that under WCS, we can then slice one year.
What are the ramificaitons for WMS? Only possible with simple 2D Coverages? No way of indicating the year?

**Comment by Mohinem** [comment]
> Hi @Mohinem,
> 
> I would give it a time axis that has year as units, and 2018 as only value. My interpretation would be that the data is 'valid' for that time (e.g. that whole year, no more precise details are known).
> 
> For the nodata value I think it is best to keep the 240. Somebody selected that on purpose, probably because the values 0 and 255 were already in use and their definition doesn't match the concept of having no data for a grid cell.

@robknapen I reimported the grassland_status_demo dataset, with the year in the format you wanted. Please check.

**Comment by robknapen** [comment]
Hi @Mohinem,

To me it looks good. Seems quite understandable to select data like this, if you know there is a time dimension involved:

```
query = '''
for $c in (grassland_status_demo)
return
  encode(
    $c[ time("2018"), X(5100000:5200000), Y(2000000:2100000) ],
    "image/tiff"
  )
'''
```

Just a small issue, in the Envelope the time UoM seems to have gotten "GridSpacing" as label, instead of "year". Perhaps that can still be adjusted?

```
"gml:Envelope": {
  "@srsName": "http://localhost:8080/rasdaman/def/crs-compound?1=http://localhost:8080/rasdaman/def/crs/OGC/0/Index1D?axis-label=\"time\"&2=http://localhost:8080/rasdaman/def/crs/EPSG/0/3035",
  "@axisLabels": "time Y X",
  "@uomLabels": "GridSpacing metre metre",
  "@srsDimension": "3",
  "gml:lowerCorner": "2018 1800000 5000000",
  "gml:upperCorner": "2018 2300000 5300000"
 }

```

**Comment by KathiSchleidt** [comment]
@pebau I see we've now moved away from temporal CRS, just enumerate the years? Is this the only way we can provide yearly data?

@sonjastndl can you check to see if this encoding still works with your code?

**Comment by pebau** [comment]
> @pebau I see we've now moved away from temporal CRS, just enumerate the years? Is this the only way we can provide yearly data?

Why do you think so? This _is_ a temporal CRS.

**Comment by pebau** [comment]
An important question is what you have in mind: is 2018 valid from Jan 01 - Dec 31? Or...? Pixel in center or in corner on the time axis? etc.

Note: this is not a technicality (those we try to hide step by step) but an intrinsic data property (which data users must understand in order to work meaningfully).

**Comment by pebau** [comment]
BTW, WMS says, time format is yyyy-MM-ddThh:mm:ss.SSSZ where the entire string is suffixed with a Z, indicating [UTC](http://en.wikipedia.org/wiki/Coordinated_Universal_Time) for the time zone. (**The WMS specification does not provide for other time zones**.)

**Comment by pebau** [comment]
Also, "WMS Servers will apply the TIME value to all temporally enabled layers in the LAYERS parameter of the GetMap request." so not good for comparing slices from different times in different layers. Source: http://opengeospatial.github.io/e-learning/wms/text/operations.html

**Comment by pebau** [comment]
> A time dimension in years would be valuable across the Copernicus Products. I'd understood that under WCS, we can then slice one year. What are the ramificaitons for WMS? Only possible with simple 2D Coverages? No way of indicating the year?

Trying to answer this one:
- If you use the TIME parameter you must select 2018, otherwise error
- if not, the server takes a default which is 2018 (communicated in the Capabilities)

**Comment by Mohinem** [comment]
A question to everyone.
Regards to time axis, what should I do that -
1. Is possible to implement
2. Satisfies the requirements by UC ?

**Comment by Mohinem** [comment]
And, what should be the date and time for the time axis ?

**Comment by pebau** [comment]
just follow the data request, literally in all detail. If that is not what a UC wants they need to update the request. That is the only chance I see that we end up with documented data.

**Comment by Mohinem** [comment]
Alright, the small demo dataset has been reimported with the time axis (called time instead of ansi), according to the specifications in https://github.com/FAIRiCUBE/data-requests/issues/10 (with the hope of getting UC approval in the foreseeable future).

**Comment by KathiSchleidt** [comment]
@pebau for most of the data requests, ALL available years have been requested. For the Biodiv use cases (NHM and WER), time is relevant (after all, we are also thinking about climate change, not climate statis ;) )

To my understanding, at present for most of the datasets, only the last year has been imported, but as a time axis has been included, the working assumption was that the additional years would be imported once we had the first time-step for all required datasets (at least I thought that this is why we have the temporal ansi CRS in addition to the spatial)

**Comment by pebau** [comment]
@KathiSchleidt yes, no problem - as long as we have indicated a time axis in the request form.

**Comment by KathiSchleidt** [comment]
@pebau see the "time coverage" field in this data request

`"time coverage": 2015, 2018`

**Comment by KathiSchleidt** [comment]
In addition, if the desired years are not indicated, I'd expect the data ingestion team to get in touch with the relavant UC partners to inquire, not just make assumptions (e.g. last year should suffice)

**Comment by Mohinem** [comment]
Alright, so update - 
1. I have fully imported the 'grassland_status_demo' dataset.
2. I have imported the dataset for year 2015 as 'grassland_status_year_2015''.
3. With the current situation, and after talking to my team many times, unfortunately I don't have a way to honour some of the UC owners' wishes - to be precise - the wish to have only one year (2018) instead of a timestamp. This is currently not possible in our systems.

**Comment by pebau** [comment]
As poor Mohit has received so much criticism I have checked the original request, it says:

T: time, axis inserted to create time series database;
Time slice - "2018-01-01T20:12:02”;
lower bound same as upper boundt now

...so with a single tie slice he has done exactly what was requested: single time slice. No time resolution indicated.

We keep the 2015 and 2018 now for single access if desired and combine both in a virtual coverage for spatio-temporal access.

**Comment by robknapen** [comment]
@Mohinem Apologies should my comments/feedback have come across as criticisms, for sure they have not been intended as such, only to help clarify and resolve questions.

**Comment by KathiSchleidt** [comment]
@pebau apologies for being so contradictory, but if I look at the data provided for this request, I see:
`Time Coverage: 2015 2018`

Mohit has now resolved this by providing 2 separate coverages for the different spatial resolutions (whereby a common pattern in naming may be valuable):
- grassland_status
- grassland_status_year_2015

However, I see closing this as premature:
- we'd agreed that when a dataset has been successfully ingested, it will then be provided to the UC responsibles for validation. That has not yet happened
- when I look at grassland_status_year_2015, the gml:Envelope indicates that this data is for "2018-01-01T08:12:02.000Z"

Once this issue has been corrected, please provide notify the relevant UC partners (in addition to Rob, the NHM colleagues) for validation!

**Comment by pebau** [comment]
@Mohinem ok, please do that:

> Once this issue has been corrected, please provide notify the relevant UC partners (in addition to Rob, the NHM colleagues) for validation!

**Comment by pebau** [comment]
>  As for why slices have different outputs in CSV or JSON, I don't know, will have to find it out.

not all formats support "null", in general mileage of features typically varies across formats - cf missing geo reference in PNG.

**Comment by KathiSchleidt** [comment]
I'd like to see what we now get for the 2 WCS requests I posted above, but have issues due to the fact that the [rasdaman certificate has expired](https://github.com/FAIRiCUBE/FAIRiCUBE-Hub-issue-tracker/issues/19)

**Comment by KathiSchleidt** [comment]
@pebau @Mohinem 
I've just looked at the new grassland_status_2015, getting ever more confused :(
- looking at the name, I'd expect 2015 data
- looking in the coverage description, I read ["2018-01-01T08:12:02.000Z", "2018-01-02T08:12:02.000Z"]

Generally, could you please fix and clean up? We currently have the following online, at least 1 is redundant (I believe one for 2015, one for 2018 would suffice

Same issues pertaining to missing rangeType and link to metadata as with other datasets :(

**Comment by pebau** [comment]
> @pebau @Mohinem I've just looked at the new grassland_status_2015, getting ever more confused :(
> Generally, could you please fix and clean up? We currently have the following online, at least 1 is redundant (I believe one for 2015, one for 2018 would suffice

Mohit will look into this.

Just a note:

> Same issues pertaining to missing rangeType and link to metadata as with other datasets :(

- a range type can never be missing, by design, think you have something else in mind.
- metadata link: we cannot link, we do not know the STAC record yet. Thought that was clear from the procedure + discussion.

**Comment by KathiSchleidt** [comment]
@pebau 
`a range type can never be missing, by design, think you have something else in mind.`
Formally, the XML is valid. Semantically, 
- name="Palette"
- definition="http://www.opengis.net/def/dataType/OGC/0/unsignedByte
- uom code="10^0"
make no sense, as discussed in #57 
When will the agreed encoding be provided?

`metadata link: we cannot link, we do not know the STAC record yet. Thought that was clear from the procedure + discussion.`
The link can be provided in the XML even if the URL doesn't yet resolve. All info long available under [Metadata for Datacubes](https://github.com/FAIRiCUBE/FAIRiCUBE-Hub-issue-tracker/issues/21)

**Comment by KathiSchleidt** [comment]
@pebau once more, please reflect on the process agreed at the meeting 2023_08_08 : Data Ingestion vs. Dataset Metadata, act accordingly

**Comment by KathiSchleidt** [comment]
@Mohinem I just checked grassland_status_virtual_coverage_index, failing badly:

ExceptionText:
Failed internal rasql query: SELECT encode((CONCAT ( MARRAY pt in [0:0] VALUES project(c0[pt[0],105000:105004,-5:-1], "3000000.0,3000000.0,3000100.0,3000100.0", "EPSG:3035", "3000000,3000000,3000100,3000100", "EPSG:3035", 10.0, 10.0) ) WITH ( MARRAY pt in [0:0] VALUES project(c1[pt[0],-200000:-199991,-80010:-80001], "3000000.0,3000000.0,3000100.0,3000100.0", "EPSG:3035", "3000000,3000000,3000100,3000100", "EPSG:3035", 10.0, 10.0) ) ALONG 0), "json" , "{\"geoReference\":{\"crs\":\"EPSG:3035\",\"bbox\":{\"xmin\":3000000,\"ymin\":3000000,\"xmax\":3000100,\"ymax\":3000100}},\"outputType\":\"GeneralGridCoverage\",\"formatParameters\":{\"enableNull\":\"false\"},\"nodata\":[]}") FROM grassland_2015_index AS c0, grassland_2018_index AS c1

**Comment by pebau** [comment]
> What's making me wonder is the value 240 in the new version where we had 0 in the old.

Looks like that is the color translation below where 240 is mapped to 0/0/0=black. That's technically ok, whether you want to see null as black is a user decision - likely, though you will want to display it as transparent and clients jump on 0 values to do that, so the translation makes sense.

> What's really making me wonder is the double appearance of 240, once as a value, once as a color encoding. The color encoding is as follows:
>     * 0: 240,240,240,255
>     * 1: 163,199,56,255
>     * 240: 0,0,0,0
>     * Rest: 0,0,0,255

No problem with the 240 occurrences - they appear in different places meaning different things. You can change the 240 grey value to anything you want without anything breaking.
